client/components/date-range/test/index.js

```diff
@@ -0,0 +1,632 @@
+			const datePicker = wrapper.find( DatePicker );
+
+			const expected = today;
+			const actual = datePicker.props().toMonth;
+
+			expect( actual ).toEqual( expected );
+		} );
+	} );
+
+	describe( 'Input elements', () => {
+		let startDate;
+		let endDate;
+		let momentStartDate;
+		let momentEndDate;
+
+		beforeEach( () => {
+			startDate = '2018-04-20';
+			endDate = '2018-05-28';
+			momentStartDate = moment( startDate );
+			momentEndDate = moment( endDate );
+		} );
+
+		test( 'should see inputs reflect date picker selection', () => {
+			const wrapper = shallow( <DateRange moment={ moment } /> );
+
+			const expectedStart = '2018-04-03';
+			const expectedEnd = '2018-04-29';
+
+			const newStartDate = moment( expectedStart );
+			const newEndDate = moment( expectedEnd );
+
+			// Select dates using API
+			wrapper.instance().onSelectDate( newStartDate );
+			wrapper.instance().onSelectDate( newEndDate );
+
+			// Force update
+			wrapper.update();
+
+			const dateRangeInputs = wrapper.find( DateRangeInputs );
+
+			expect( dateRangeInputs.props() ).toEqual(
+				expect.objectContaining( {
+					startDateValue: dateToLocalString( newStartDate ),
+					endDateValue: dateToLocalString( newEndDate ),
+				} )
+			);
+		} );
+
+		test( 'should update start date selection on start date input blur event', () => {
+			const wrapper = shallow( <DateRange moment={ moment } /> );
+
+			wrapper.instance().handleInputBlur( '04/20/2018', 'Start' );
+
+			expect( wrapper.state().startDate.format( 'L' ) ).toEqual( momentStartDate.format( 'L' ) );
+		} );
+
+		test( 'should update end date selection on end date input blur event', () => {
+			const wrapper = shallow( <DateRange moment={ moment } /> );
+
+			wrapper.instance().handleInputBlur( '05/28/2018', 'End' );
+
+			expect( dateToLocalString( wrapper.state().endDate ) ).toEqual(
+				dateToLocalString( momentEndDate )
+			);
+		} );
+
+		test( 'should not update date selection on input change event', () => {
+			const wrapper = shallow( <DateRange moment={ moment } /> );
+
+			wrapper.instance().handleInputChange( '05/28/2018', 'End' );
+
+			expect( dateToLocalString( wrapper.state().endDate ) ).toEqual(
+				dateToLocalString( fixedEndDate )
+			);
+		} );
+
+		test( 'should update `textInput*` state on input change event', () => {
+			const wrapper = shallow( <DateRange moment={ moment } /> );
+
+			wrapper.instance().handleInputChange( '05/28/2018', 'End' );
+
+			expect( wrapper.state().textInputEndDate ).toEqual( dateToLocalString( momentEndDate ) );
+		} );
+
+		test( 'should not update either start or end date selection if the new input date value is the same as that stored in state', () => {
+			const wrapper = shallow(
+				<DateRange
+					selectedStartDate={ momentStartDate }
+					selectedEndDate={ momentEndDate }
+					moment={ moment }
+				/>
+			);
+
+			wrapper.instance().handleInputBlur( '04/20/2018', 'Start' );
```

Comment:
Just a thought here - no need to update.
When reading through tests in a large test file I like to see inputs defined locally to the tests where possible or for constants to be used so that I don't need to scroll up and down the file for context.
Here for example, I'd have to remind myself of the significance of `'04/20/2018'` in relation to `momentStartDate` by scrolling up (which interrupts my reading of the test cases).

We can be more redundant with tests as their performance isn't as important as it is in a product so redefining moments and such is fine if it helps make the intent of the test case more clear:

```js
const startDate = '04/20/2018';
const startDateMoment = moment( startDate );
const wrapper = shallow(
  <DateRange
    selectedStartDate={ momentStartDate }
    selectedEndDate={ momentEndDate }
    moment={ moment }
  />
);

wrapper.instance().handleInputBlur( startDate, 'Start' );
```

Just wanna hammer home that this isn't a change request as whats here is great, just a personal thought/opinion :)

Reply:
I'm glad you raised this. I've been thinking about this and having done a lot of reading about testing best practices recently I'm convinced that we should avoid abstracting tests as much as possible in favour of clarity and simplicity.

I'll look into a refactor of this - shouldn't take long.

Reply:
I totally agree with this. Having done a lot more reading about testing best practices, I'm also convinced that it's better to have redundancy in rests and be clearer.

> no need to update.

Agree. I don't think I will refactor these now but I will keep in mind for next time.